transaction/core/src/tx.rs

```diff
@@ -262,26 +268,92 @@ impl TxOut {
+            Ok(MemoPayload::default())
+        })
+        .map_err(|err| match err {
+            NewTxError::Amount(err) => err,
+            NewTxError::Memo(_) => unreachable!(),
+        })
+    }
+
+    /// Creates a TxOut that sends `value` to `recipient`, with a custom memo
+    /// attached. The memo is produced by a callback function which is
+    /// passed the value and tx_public_key.
+    ///
+    /// # Arguments
+    /// * `value` - Value of the output.
+    /// * `recipient` - Recipient's address.
+    /// * `tx_private_key` - The transaction's private key
+    /// * `hint` - Encrypted Fog hint.
+    /// * `memo_fn` - A callback taking tx_public_key, which produces a
+    ///   MemoPayload, or a String error
+    pub fn new_with_memo(
+        value: u64,
+        recipient: &PublicAddress,
+        tx_private_key: &RistrettoPrivate,
+        hint: EncryptedFogHint,
+        memo_fn: impl FnOnce(&RistrettoPublic) -> Result<MemoPayload, String>,
+    ) -> Result<Self, NewTxError> {
         let target_key = create_onetime_public_key(tx_private_key, recipient).into();
-        let public_key = create_tx_public_key(tx_private_key, recipient.spend_public_key()).into();
+        let public_key = create_tx_public_key(tx_private_key, recipient.spend_public_key());
+
+        let shared_secret = create_shared_secret(recipient.view_public_key(), tx_private_key);
 
-        let amount = {
-            let shared_secret = create_shared_secret(recipient.view_public_key(), tx_private_key);
-            Amount::new(value, &shared_secret)
-        }?;
+        let amount = Amount::new(value, &shared_secret)?;
+        let memo = memo_fn(&public_key).map_err(NewTxError::Memo)?;
+        let e_memo = memo.encrypt(&shared_secret);
 
         Ok(TxOut {
             amount,
             target_key,
-            public_key,
+            public_key: public_key.into(),
             e_fog_hint: hint,
+            e_memo: Some(e_memo),
         })
     }
 
-    /// Blake2B256 hash of his TxOut.
+    /// A merlin-based hash of this TxOut.
     pub fn hash(&self) -> Hash {
         self.digest32::<MerlinTranscript>(b"mobilecoin-txout")
     }
+
+    /// Try to decrypt the e_memo field, using the TxOut shared secret.
+    ///
+    /// This function is backwards-compatible in the following sense:
+    /// - If self.e_memo is empty (a TxOut from before this field was added), we
+    ///   return MemoPayload corresponding to "unused memo".
+    /// - If self.e_memo is present, we use MemoPayload::try_decrypt. This
+    ///   succeeds unless the e_memo has an invalid length.
+    ///
+    /// This function only returns an error if e_memo has a length other than 0
+    /// or 46.
+    ///
+    /// Note that the results of this function call are unauthenticated.
+    pub fn try_decrypt_memo(
+        &self,
+        tx_out_shared_secret: &RistrettoPublic,
+    ) -> Result<MemoPayload, LengthError> {
```

Comment:
Looks like this will never fail, why a result?

Reply:
will fix